### PR TITLE
NAS-115279 / 22.02.1 / Normalize REST resources for jobs that use pipes:

### PR DESCRIPTION
**Pull request by themylogin**
* Allow specifying `?download=0` to avoid providing a pipe to the method (useful for `pool.dataset.export_keys`)
* Allow uploading files via REST resources
* Document how to work with such resources
* Add integration tests

### CODE REVIEW

**Comment by themylogin** [comment]
time 11:30

File: src/middlewared/middlewared/restful.py [hunk 1]
```diff
@@ -529,30 +551,109 @@ async def do(self, http_method, req, resp, **kwargs):
...
+                    return resp
+
+        if upload_pipe and download_pipe:
+            method_kwargs['pipes'] = Pipes(input=upload_pipe, output=download_pipe)
+        elif upload_pipe:
+            method_kwargs['pipes'] = Pipes(input=upload_pipe)
+        elif download_pipe:
+            method_kwargs['pipes'] = Pipes(output=download_pipe)
+
+        if True:
```
Comment (Qubad786) [inline src/middlewared/middlewared/restful.py:643]: Is this required ?

File: src/middlewared/middlewared/service.py [hunk 2]
```diff
@@ -1624,14 +1624,22 @@ def job_abort(self, id):
...
+        if service._config.private is True:
+            if not (target == 'REST' and name == 'resttest'):
+                return False
+
+        if target == 'CLI' and service._config.cli_private:
+            return False
+
+        return True
+
+    @accepts(Str('target', enum=['WS', 'CLI', 'REST'], default='WS'))
```
Comment (Qubad786) [inline src/middlewared/middlewared/service.py:1637]: Should we add `returns` decorator here?

File: src/middlewared/middlewared/restful.py (hunk 1 above)
Comment (themylogin) [reply src/middlewared/middlewared/restful.py:643]: I keep it for smaller and more readable diff. I'll remove this and unindent the code once this PR once it is reviewed and aproved

**Comment by themylogin** [comment]
https://github.com/truenas/midcli/pull/53 must be merged when merging this PR

**Review by Qubad786 — APPROVED** [review]
LGTM!

:shipit: after removing `if True:` ;)

**Comment by themylogin** [comment]
backport

**Review by rick-mesta — APPROVED** [review]

**Pull request merged**